doc/cookbook/current-route/CurrentRoute.lhs

```diff
@@ -0,0 +1,391 @@
+# How to get current route in servant-server?
+
+This tries to be an answer to the [StackOverflow question](https://stackoverflow.com/questions/41538438/haskell-servant-get-current-route-url-from-handler).
+There are two questions in one:
+
+1. how to get current Request or URL?
+2. how to get current "route"?
+
+Note, that URL (e.g. `/route12/42`) is different than route
+(e.g. `"route12" :> Capture "id" Int :> Get '[JSON] Int).
+Let's see how we can solve both of these questions, right after a
+short language pragma and import section.
+
+```haskell
+{-# LANGUAGE ConstraintKinds         #-}
+{-# LANGUAGE DataKinds               #-}
+{-# LANGUAGE DeriveGeneric           #-}
+{-# LANGUAGE FlexibleContexts        #-}
+{-# LANGUAGE FlexibleInstances       #-}
+{-# LANGUAGE MultiParamTypeClasses   #-}
+{-# LANGUAGE OverloadedStrings       #-}
+{-# LANGUAGE RankNTypes              #-}
+{-# LANGUAGE ScopedTypeVariables     #-}
+{-# LANGUAGE TypeFamilies            #-}
+{-# LANGUAGE TypeOperators           #-}
+{-# LANGUAGE UndecidableInstances    #-}
+{-# LANGUAGE UndecidableSuperClasses #-}
+{-# OPTIONS_GHC -Wno-orphans         #-}
+module Main where
+
+import Data.Maybe             (fromMaybe)
+import Control.Monad.IO.Class (liftIO)
+import System.Environment     (getArgs)
+import GHC.Generics           (to, from, M1 (..), K1 (..), (:*:) (..))
+
+-- for "unsafe" vault key creation
+import System.IO.Unsafe (unsafePerformIO)
+
+import qualified Data.ByteString.Char8    as BS8
+import qualified Data.Vault.Lazy          as V
+import qualified Network.Wai              as Wai
+import qualified Network.Wai.Handler.Warp as Warp
+
+import Servant
+import Servant.API.Generic 
+import Servant.Server.Generic
+import Servant.Server.Internal.RoutingApplication (passToServer)
+```
+
+## How to get current `Request` object or URL
+
+Passing current `WAI` `Request` to the handler is actually quite easy.
+This is "lazy" approach, we ask for "everything" in the request,
+and we have to be careful in the handler (e.g. we cannot touch `requestBody`).
+Also this "combinator" ties implementation to the `wai` server implementation,
+which is an implementation detail
+(nothing else in `servant-server` exposes `wai` internals, except of `Raw`).
+
+The idea is to make `Server (Wai.Request :> api) = Wai.Request -> Server api`.
+If we imagine for a second that we have such functionality in place,
+we can write, using `Servant.API.Generic` (see "Using generics" cookbook recipe):
```

Comment:
Would be great to link to the recipe

Reply:
How the intra-cookbook linking works?

Reply:
Good question. I don't know if there's a way to give relative links. If not, just link to the stable/latest version, that's fine.